File: iosApp/iosApp/Ui/Weather/WeatherItem.swift
//
//  WeatherItem.swift
//  iosApp
//
//  Created by openobject2 on 9/16/25.
//

import SwiftUI
import Shared

struct CurrentWeatherInfo: View {
    let state: WeatherState
    private let temp: String
    private let tempMax: String
    private let tempMin: String
    
    init(state: WeatherState) {
        self.state = state
        temp = "\(Int(state.curWeather!.main.temp.rounded()))℃"
        tempMax = "\(Int(state.curWeather!.main.tempMax.rounded()))℃"
        tempMin = "\(Int(state.curWeather!.main.tempMin.rounded()))℃"
    }
    
    var body: some View {
        VStack(alignment: .center, spacing: 16) {
            Text(state.coordinates?.first?.localName?.ko ?? state.curWeather!.name)
                .foregroundStyle(Color.white)
                .font(Font.system(size: 24))
            
            Text(temp)
                .foregroundStyle(Color.white)
                .fontWeight(.light)
                .font(.system(size: 48))
            
            Text(state.curWeather!.weather.first!.description_)
                .foregroundStyle(Color.white)
                .font(.system(size: 16))
            
            Text("최고 \(tempMax)\n최저 \(tempMin)")
                .foregroundStyle(Color.white)
                .font(.system(size: 14))
        }
        .fixedSize(horizontal: false, vertical: true)
    }
}

struct CitySearchDropdown: View {
    @State private var selectedCity = 0
    let isExpanded: Bool
    let availableCities: [CityCoordinate]
    let onCityClick: (Int) -> Void
    let onDismiss: () -> Void
    
    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            ForEach(availableCities.indices, id: \.self) { idx in
                Button(action: { onCityClick(idx) }) {
                    Text(availableCities[idx].localName?.ko ?? "")
                        .tag(idx)
                        .foregroundColor(.black)
                        .padding(.vertical, 12)
                        .padding(.horizontal, 16)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
        .fixedSize(horizontal: false, vertical: true)
        .background(Color.white)
        .cornerRadius(5)
        .shadow(radius: 5)
        .padding(.horizontal)
        .onDisappear(perform: onDismiss)
    }
}

struct OtherWeatherInfo: View {
    let otherItems: [OtherInfoUiModel]
    private let columns: [GridItem] = [
        GridItem(.flexible(), spacing: 4),
        GridItem(.flexible(), spacing: 4),
    ]
    
    var body: some View {
        LazyVGrid(columns: columns, spacing: 4) {
            ForEach(otherItems, id: \.title) { item in
                WeatherItem(title: item.title, content: item.content)
            }
        }
        .padding(.horizontal, 16)
        .fixedSize(horizontal: false, vertical: true)
    }
}

struct WeatherItem: View {
    let title: String
    let content: String
    
    var body: some View {
        VStack(alignment: .center) {
            Text(title).font(.footnote)
            Spacer(minLength: 10)
            getUnitAttributedText(prev: content, unitSize: 16)
        }
        .frame(maxWidth: .infinity)
        .foregroundStyle(.white)
        .padding(10)
        .background(
            RoundedRectangle(cornerRadius: 15).fill(Color(red: 0.52, green: 0.52, blue: 0.52, opacity: 0.52))
        )
    }
}

File: composeApp/src/androidMain/kotlin/com/dosystem/todo/ui/weather/WeatherItem.kt
package com.dosystem.todo.ui.weather

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.dosystem.todo.data.model.geo.CityCoordinate
import com.dosystem.todo.getUnitAnnotatedString
import com.dosystem.todo.viewmodel.OtherInfoUiModel
import com.dosystem.todo.viewmodel.WeatherState
import kotlin.math.roundToInt

@Composable
fun CurrentWeatherInfo(
    modifier: Modifier = Modifier,
    state: WeatherState
) {
    val temp = "${state.curWeather!!.main.temp.roundToInt()}℃"
    val tempMax = "${state.curWeather!!.main.tempMax.roundToInt()}℃"
    val tempMin = "${state.curWeather!!.main.tempMin.roundToInt()}℃"

    Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
        Text(
            text = state.coordinates?.firstOrNull()?.localName?.ko
                ?: state.curWeather!!.name,
            color = Color.White,
            fontSize = 24.sp,
            textAlign = TextAlign.Center,
            modifier = modifier.fillMaxWidth()
        )
        Text(
            text = temp,
            color = Color.White,
            fontSize = 48.sp,
            fontWeight = FontWeight.Light,
            textAlign = TextAlign.Center,
            modifier = modifier.fillMaxWidth()
        )
        Text(
            text = state.curWeather!!.weather.first().description,
            color = Color.White,
            fontSize = 16.sp,
            textAlign = TextAlign.Center,
            modifier = modifier.fillMaxWidth()
        )
        Text(
            text = "최고 $tempMax\n최저 $tempMin",
            color = Color.White,
            fontSize = 14.sp,
            textAlign = TextAlign.Center,
            modifier = modifier.fillMaxWidth()
        )
    }
}

@Composable
fun OtherWeatherInfo(
    modifier: Modifier = Modifier,
    otherItems: List<OtherInfoUiModel>
) {
    Column(
        verticalArrangement = Arrangement.spacedBy(4.dp),
        modifier = modifier.fillMaxWidth()
    ) {
        for (i in 0 until otherItems.size / 2) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(4.dp),
                modifier = modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp)
            ) {
                WeatherItem(
                    parentModifier = modifier.weight(1f),
                    title = otherItems[i * 2].title,
                    content = otherItems[i * 2].content
                )
                WeatherItem(
                    parentModifier = modifier.weight(1f),
                    title = otherItems[i * 2 + 1].title,
                    content = otherItems[i * 2 + 1].content
                )
            }
        }
    }

}

@Composable
fun CitySearchDropdown(
    modifier: Modifier = Modifier,
    expanded: Boolean,
    availableCities: List<CityCoordinate>,
    onDismiss: () -> Unit,
    onCityClick: (Int) -> Unit
) {
    Box {
        DropdownMenu(
            modifier = modifier
                .fillMaxWidth()
                .background(Color.White),
            expanded = expanded,
            onDismissRequest = onDismiss
        ) {
            for (i in availableCities.indices) {
                DropdownMenuItem(
                    modifier = modifier,
                    text = {
                        Text(text = availableCities[i].localName?.ko ?: "")
                    },
                    onClick = { onCityClick(i) }
                )
            }
        }
    }
}

@Composable
fun WeatherItem(
    modifier: Modifier = Modifier,
    parentModifier: Modifier,
    title: String,
    content: String
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = parentModifier
            .background(Color(0x85858585), RoundedCornerShape(15.dp))
            .padding(20.dp)
    ) {
        Text(
            text = title,
            fontSize = 12.sp,
            color = Color.White
        )
        Spacer(modifier.height(10.dp))

        Text(
            text = getUnitAnnotatedString(
                prev = content,
                unitSize = 16.sp
            ),
            fontSize = 20.sp,
            color = Color.White
        )
    }
}